[CELEBORN-955] Re-run Spark Stage for Celeborn Shuffle Fetch Failure

ErikFang opened the pull request:
### What changes were proposed in this pull request?
Currently, Celeborn uses replication to handle shuffle data lost for celeborn shuffle reader, this PR implements an alternative solution by Spark stage resubmission.

Design doc:
https://docs.google.com/document/d/1dkG6fww3g99VAb1wkphNlUES_MpngVPNg8601chmVp8/edit

### Why are the changes needed?
Spark stage resubmission uses less resources compared with replication, and some Celeborn users are also asking for it


### Does this PR introduce _any_ user-facing change?
a new config celeborn.client.fetch.throwsFetchFailure is introduced to enable this feature


### How was this patch tested?
two UTs are attached, and we also tested it in Ant Group's Dev spark cluster

ErikFang commented:
In my local spark 3.2 build, below code snippet is added to trigger DAGScheduler to clean up mapOutputTracker status
```diff
diff --git a/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala b/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala
index f97c426348..fef6013d98 100644
--- a/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala
+++ b/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala
@@ -1844,6 +1844,12 @@ private[spark] class DAGScheduler(
             }
           }

+          // ensure retry of all MapTasks of resubmitted stage attempt
+          // TODO: change to ShuffleManagerIndicator.isRemoteShuffle
+          if (failureMessage.contains("Celeborn FetchFailure: ")) {
+            mapOutputTracker.unregisterAllMapAndMergeOutput(shuffleId)
+          }
+
           if (failedStage.rdd.isBarrier()) {
             failedStage match {
               case failedMapStage: ShuffleMapStage =>
```
this is also the reason why CelebornFetchFailureSuite fails with community spark in this PR
will update a commit soon to include the RPC related changes for LifecycleManager to clean up mapOutputTracker

RexXiong commented on client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:
```diff
@@ -556,6 +562,28 @@ class LifecycleManager(val appUniqueId: String, val conf: CelebornConf) extends
+      shuffleIdMapping.put(appShuffleId, shuffleIds :+ newShuffleId)
+    }
+
+    val shuffleId = shuffleIdMapping.get(appShuffleId).last
```
For Shuffle unregister/clean uselessShuffleId 
1. When register a new shuffleId with same appShuffleId we can clear all out-dated info in lifecycleManager, such as shuffleIdMapping(appShuffleId->shuffleId), shuffleId related info...
2. When shuffleId unregister, LifecycleManager alse can clear the useless shuffleId in shuffleIdMapping.

ErikFang replied:
shuffle clean up has been added:
1. In LifecycleManager, unregisterShuffle() is invoked when handing fetch failure
2. new function unregisterAppShuffle() is added to LifecycleManager what will be called by SparkShuffleManager.unregisterShuffle() with appShuffleId
3. regarding executor, shuffleIdTracker is added to track and cleanup related shuffle id in shuffleClient

SteNicholas commented on common/src/main/proto/TransportMessages.proto:
```diff
@@ -278,6 +282,26 @@ message PbGetReducerFileGroupResponse {
   repeated int32 partitionIds = 4;
 }
 
+message PbGetShuffleId {
```
Does the `TransportMessage#getParsedPayload` need to add these message?

ErikFang replied:
can you help to explain what kind of message needs to be added to TransportMessage#getParsedPayload?

SteNicholas replied:
@ErikFang, all PB messages should use `TransportMessage#getParsedPayload` to parse the payload.

SteNicholas commented:
@ErikFang, could you use `dev/reformat` to format the changes?

ErikFang commented on client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:
```diff
@@ -556,6 +562,28 @@ class LifecycleManager(val appUniqueId: String, val conf: CelebornConf) extends
     commitManager.handleGetReducerFileGroup(context, shuffleId)
   }
 
+  private def handleGetShuffleId(
```
> Another option:
> 
> 1. For writer, Add new field appShuffleId to registerShuffleRequest and generate uniqueShuffleId(used as celeborn shuffleId) at client side(spark), then lifecycleManager doesn't have to compute shuffleId according to spark stage rerun-logic, and writer needn't introduce a new rpc, btw shuffleIdMapping still need in lifecyceManager.

This is good, I've ,made the change locally, will update the PR soon

> 2. For Reader, shuffleClient can directly get latest shuffleId related, this way we don't need the getShuffleId rpc, but we need change the shuffleId meaning in GetReducerFileGroup to appShuffleId, and response latest shuffle id in GetReducerFileGroupResponse @ErikFang @waitinfuture

I checked the code, but got problem with "response latest shuffle id in GetReducerFileGroupResponse"
the call hierarchy is 
```
ShuffleClientImpl.readPartition(int appShuffleId, ....)
            |
ShuffleClientImpl.loadFileGroup(int appShuffleId, ...)
            |
ShuffleClientImpl.updateFileGroup(int appShuffleId, ...)
            |
ShuffleClientImpl.loadFileGroupInternal(int appShuffleId)
```
and reduceFileGroupsMap is of type
```
 protected final Map<Integer, ReduceFileGroups> reduceFileGroupsMap =
      JavaUtils.newConcurrentHashMap();
```
, which maps shuffleId -> ReduceFileGroups and updates with
`reduceFileGroupsMap.computeIfAbsent(shuffleId, (id) -> loadFileGroupInternal(shuffleId))`
If shuffleId is obtained from GetReducerFileGroupResponse, then current computeIfAbsent() logic will be changed and every updateFileGroup() call sends a GetReducerFileGroup RPC to lifecycleManager, which may impact performace

How about keeping the get shuffleId RPC, or any advice? @RexXiong @waitinfuture

waitinfuture commented on client/src/main/java/org/apache/celeborn/client/ShuffleClient.java:
```diff
@@ -200,7 +247,16 @@ public abstract PartitionLocation registerMapPartitionTask(
+
+  public abstract int getShuffleId(int appShuffleId);
+
+  public abstract boolean reportShuffleFetchFailure(int appShuffleId, int uniqueShuffleId);
```
`uniqueShuffleId` -> `celebornShuffleId`

waitinfuture commented on client/src/main/java/org/apache/celeborn/client/ShuffleClient.java:
```diff
@@ -200,7 +247,16 @@ public abstract PartitionLocation registerMapPartitionTask(
 
   public abstract ConcurrentHashMap<Integer, PartitionLocation> getPartitionLocation(
-      int shuffleId, int numMappers, int numPartitions);
+      int shuffleId, int appShuffleId, int numMappers, int numPartitions);
```
Better to unify the variable names, for example `sparkShuffleId`/`appShuffleId` -> `appShuffleId`, `uniqueShuffleId`/`celebornShuffleId` -> `celebornShuffleId`

ErikFang replied:
I suggest to follow below rules for variable names:
1. in spark package, sparkShuffleId/shuffleId
2. In celeborn package, appShuffleId/shuffleId

It doesn't make sense for celebornShuffleId in celeborn package, for instance, lots of shuffleId variable name is used in LifecycleManager
but it is ok for me to combine both sparkShuffleId/appShuffleId to appShuffleId in spark/celeborn packages if you prefer

waitinfuture replied:
I do prefer unify sparkShuffleId/appShuffleId to appShuffleId is better

waitinfuture commented on client-spark/common/src/main/java/org/apache/spark/shuffle/celeborn/SortBasedPusher.java:
```diff
@@ -61,6 +61,7 @@ public class SortBasedPusher extends MemoryConsumer {
   private final int pushBufferMaxSize;
   private final long pushSortMemoryThreshold;
   private final int shuffleId;
+  private final int sparkShuffleId;
```
unused?

ErikFang replied:
no, this is a bug, it is fixed in latest commit

waitinfuture commented on client/src/main/java/org/apache/celeborn/client/write/DataPushQueue.java:
```diff
@@ -48,6 +48,7 @@ public class DataPushQueue {
   private final DataPusher dataPusher;
   private final int maxInFlightPerWorker;
   private final int shuffleId;
+  private final int appShuffleId;
```
unused?

ErikFang replied:
no, this is a bug, it is fixed in latest commit

SteNicholas commented on client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SparkUtils.java:
```diff
@@ -92,6 +101,25 @@ public static String appUniqueId(SparkContext context) {
         .getOrElse(context::applicationId);
   }
 
+  public static int celebornShuffleId(ShuffleClient client, CelebornShuffleHandle<?, ? ,?> handle, TaskContext context, Boolean isWriter) {
```
```suggestion
  public static int shuffleId(ShuffleClient client, CelebornShuffleHandle<?, ? ,?> handle, TaskContext context, Boolean isWriter) {
```
IMO, no need to emphasize that the return value is celeborn shuffle id.

waitinfuture replied:
We need to distinguish from appShuffleId and celebornShuffleId, as this PR introduces a mapping from spark's shuffleId to celeborn's shuffleId.

SteNicholas commented on client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SparkUtils.java:
```diff
@@ -46,6 +50,11 @@
+  // copied from DAGScheduler.DEFAULT_MAX_CONSECUTIVE_STAGE_ATTEMPTS because can't access it directly
+  public static final String MAX_CONSECUTIVE_STAGE_ATTEMPTS = "spark.stage.maxConsecutiveAttempts";
+  public static final int DEFAULT_MAX_CONSECUTIVE_STAGE_ATTEMPTS = DAGScheduler.DEFAULT_MAX_CONSECUTIVE_STAGE_ATTEMPTS();
+  public static final String FetchFailureErrorMsg = "Celeborn FetchFailure! ";
```
```suggestion
  public static final String FETCH_FAILURE_ERROR_MSG = "Celeborn FetchFailure! ";
```

waitinfuture commented on common/src/main/scala/org/apache/celeborn/common/protocol/message/ControlMessages.scala:
```diff
@@ -948,6 +993,28 @@ object ControlMessages extends Logging {
           partitionIds)
 
+      case GET_SHUFFLE_ID_VALUE =>
+        val pbGetShuffleId = PbGetShuffleId.parseFrom(message.getPayload)
```
As @SteNicholas suggested, should use `TransportMessage#getParsedPayload`

ErikFang replied:
```
      case GET_SHUFFLE_ID_VALUE =>
        message.getParsedPayload()
//        PbGetShuffleId.parseFrom(message.getPayload)
```
so...is my understanding correct?

waitinfuture replied:
yes

SteNicholas commented on client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SparkShuffleManager.java:
```diff
@@ -37,6 +38,18 @@
 import org.apache.celeborn.reflect.DynMethods;
 
+/*
+  In order to support Spark Stage resubmit with ShuffleReader FetchFails, Celeborn shuffleId has to be distinguished
```
```suggestion
  In order to support Spark Stage rerun with ShuffleReader fetch failure, Celeborn shuffleId must be distinguished 
```

SteNicholas commented on common/src/main/proto/TransportMessages.proto:
```diff
@@ -76,6 +76,10 @@ enum MessageType {
   CHECK_WORKERS_AVAILABLE = 53;
   CHECK_WORKERS_AVAILABLE_RESPONSE = 54;
   REMOVE_WORKERS_UNAVAILABLE_INFO = 55;
+  REPORT_SHUFFLE_FETCH_FAILURE = 62;
```
Please rebase main branch and update this value.

SteNicholas reviewed (changes requested):
@ErikFang, I left some comments for the updates. PTAL.

SteNicholas commented:
@ErikFang, could you rebase the latest main branch to resolve conflicts for this pull request?

ErikFang commented:
I've update the PR to reformat & rebase

waitinfuture commented:
> I've update the PR to reformat & rebase

Hi @ErikFang , please run ./dev/reformat to auto format the code style.

ErikFang commented on client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:
```diff
@@ -556,6 +566,59 @@ class LifecycleManager(val appUniqueId: String, val conf: CelebornConf) extends
+          shuffleIds += ((newShuffleId, true))
+          shuffleIdMapping.put(appShuffleId, shuffleIds)
+        }
+      }
```
this logic has been changed and no isShuffleWriter now

ErikFang commented on common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:
```diff
@@ -3077,6 +3078,15 @@ object CelebornConf extends Logging {
       .createWithDefault(3)
 
+  val CLIENT_FETCH_THROWS_FETCH_FAILURE: ConfigEntry[Boolean] =
+    buildConf("celeborn.client.fetch.throwsFetchFailure")
```
added it

ErikFang commented:
update a new commit to address comments related to code style/comment/variable name

waitinfuture commented:
Hi @ErikFang , please add license header for new added files

mridulm commented on client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SparkShuffleManager.java:
```diff
@@ -38,6 +38,16 @@
+ throws {@link FetchFailedException}, Spark DAGScheduler resubmits the failed ResultStage and corresponding ShuffleMapStage
+ , but Celeborn can't differentiate shuffle data from previous failed/resubmitted ShuffleMapStage with the same shuffleId.
+ Current solution takes Stage retry in account, and set
+         celeborn shuffleId = spark shuffleId * maxStageAttemptsNum + stageAttemptsId
```
I probably mentioned this in the discussion in dev list - we cannot assume `spark shuffleId * maxStageAttemptsNum + stageAttemptsId` does not lead to conflicts.

With shuffle id reuse across jobs, this will result in conflicts.

ErikFang replied:
Mridul, as I explained in maillist, spark shuffle id is a static concept bounded to the shuffle dependency, aka, the shuffle data. I prefer not to call it "shuffle id reuse", since it is not the same id pointing to different data, it is just THE shuffle data's id. I don't think there will be any conflict.
If spark shuffle data 0 needs to be accessed, no matter it will be submitted to DAGScheduler soon, or it had been computed in previous job, spark shuffle id 0 should be used. In second case, spark read the shuffle data directly if the data doesn't need to be re-computed. you can search 'spark_shuffleId 0' in logs pasted in my mail to see that, shuffle data 0 is written once and read twice with spark shuffle id 0/celeborn shuffle id 0

ErikFang replied:
> I probably mentioned this in the discussion in dev list - we cannot assume `spark shuffleId * maxStageAttemptsNum + stageAttemptsId` does not lead to conflicts.
> 
> With shuffle id reuse across jobs, this will result in conflicts.

Mridul, as I explained in maillist, spark shuffle id is a static concept bounded to the shuffle dependency, aka, the shuffle data. I prefer not to call it "shuffle id reuse", since it is not the same id pointing to different data, it is just THE shuffle data's id. I don't think there will be any conflict.
If spark shuffle data 0 needs to be accessed, no matter it will be submitted to DAGScheduler soon, or it had been computed in previous job, spark shuffle id 0 should be used. In second case, spark read the shuffle data directly if the data doesn't need to be re-computed. you can search 'spark_shuffleId 0' in logs pasted in my mail to see that, shuffle data 0 is written once and read twice with spark shuffle id 0/celeborn shuffle id 0

mridulm replied:
Spark uses shuffle id, stage reexecution, job reexecution (particularly for sql) in a specific way, and that does not change due to Celeborn being used.

In this specific context, the assumption being made is, for a given Spark shuffle id, there are upto maxStageAttemptNums stage attempts possible - if there is more, the generated id will conflict with some other Spark shuffle id's Celeborn shuffle id - given this, the question is, can this happen - and as I have illustrated before - it can, and does actually.

Please let me know if I am missing something etching or can help clarify

mridulm replied:
To add, `spark.stage.maxConsecutiveAttempts` is for a specific stage, not shuffle id - so total number of times a particular shuffle id is compted by stage attempts has no bounds (across jobs)

ErikFang replied:
Thanks for the clarification, you are correct.
I feel the shuffle id mapping by (maxStageAttemptsNum, stageAttemptsId) is a bad idea now
The bounded shuffle id range enforced by maxStageAttemptsNum conflicts with the unbounded spark shuffle id range generated by spark stage re-run, which may happens when shuffle data gets recomputed with several round of shuffle fetch failure

Let me review the solution of singleton shuffle id generation in LifecycleMananger, and keep track of the correct shuffleId with
`private val shuffleIdMapping = JavaUtils.newConcurrentHashMap[Int, scala.collection.mutable.LinkedHashMap[Int, Boolean]]()`

mridulm replied:
Please do let me know if I can give help or give clarity ... I don't understand Celeborn well yet (particularly given it supports multiple engines), but I can help explain how Spark might behave for your proposals !

ErikFang replied:
basic idea of shuffle id generation in LifecycleManager:
1. shuffle writer sends app shuffle id to LifecycleManager, which may create new shuffle id for writing
- If the app shuffle id is not recorded in shuffleIdMapping, then generates a new shuffle id with AtomicInteger, puts appShuffleId -> (shuffleId -> true) in shuffleIdMapping, returns shuffleId to shuffle writer 
- If the app shuffle id is mapped to (shuffleId -> true), returns shuffleId to shuffle writer 
- If the app shuffle id is mapped to (shuffleId -> false), generates a new (shuffleId' -> true) to put into shuffleIdMapping, returns shuffleId' to shuffle writer // fetch failure happened

2. shuffle reader sends app shuffle id to LifecycleManager, gets latest shuffle id associated with app shuffle id 

3. during FetchFailure, shuffle reader sends ReportFetchFailure with (app shuffleId, shuffle id) to LifecycleManager, updates shuffleIdMapping to appShuffleId -> (shuffleId -> false)

@mridulm 
Do you see any issue here?

mridulm replied:
To clarify, the incoming request to LifecycleManager will include app-shuffle-id, stage-id, stage-attempt-id ?
If yes, agree, this should address all the concerns I was having.

One small nit, if we can generate this at driver and ship the value to executor (so that we can avoid the invocation from executor to driver to fetch this id), that would be great - I have not thought through how we can do this, but adding as a comment in case there is an existing solution we can leverage !

ErikFang replied:
It is definitely a good idea to use app-shuffle-id + stage-id + stage-attempt-id as an identifier for one execution of shuffle computing, which maps to a celeborn shuffle id. let's call it appShuffleIdentifier
`val appShuffleIdentifier = s"$appShuffleId-$stageId-$stageAttemptId"`

the solutions is updated as:
define 
```
  // app shuffle id -> LinkedHashMap of (app shuffle identifier, (shuffle id, fetch status))
  private val shuffleIdMapping = JavaUtils.newConcurrentHashMap[Int, scala.collection.mutable.LinkedHashMap[String, (Int, Boolean)]]()
```
1. shuffle writer sends (appShuffleId, appShuffleIdentifier, isWriter=true) to LifecycleManager

- If appShuffleId or appShuffleIdentifier is not recorded, then generates a new shuffle id with AtomicInteger, saves mapping of appShuffleId -> appShuffleIdentifier -> (shuffleId, true) and return to writer
- if appShuffleIdentifier is found, return shuffleId to writer

2. shuffle reader sends (appShuffleId, null, isWriter=false) to LifecycleManager

- find the latest finished(all shuffle write task is done in celeborn) shuffle id associated with appShuffleId, return to reader

3. during FetchFailure, shuffle reader sends ReportFetchFailure with (app shuffleId, shuffle id) to LifecycleManager, cleans status on MapOutputTracker (exactly once) and updates fetch status of the shuffle id to false, and does nothing if fetch status is already false

Regarding 2, reader needs to get the latest finished shuffleId to prevent reader from a previous job to incorrectly get the running resubmitted ShuffleMapStage's shuffleId (that is unfinished) and reports fetch failure mistakenly

mridulm commented on client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SparkUtils.java:
```diff
@@ -92,6 +100,25 @@ public static String appUniqueId(SparkContext context) {
+      int appShuffleId = handle.shuffleId();
+      int maxAttemptNum = handle.maxStageAttemptsNum();
+      int attemptId = context.stageAttemptNumber();
+      return appShuffleId * maxAttemptNum + attemptId;
```
If this method is trying to create a new celeborn shuffle id always, why not simply use an `AtomicInteger` to generate a new id each time ?

ErikFang replied:
In current implementation, SparkUtils.celebornShuffleId(...) is called in shuffleWriter that is located in Spark executor, then there is no singleton AtomicInteger as a Id generator across the executors
We can use a single Id generator unless the id generation logic moves to Lifecyclemanger

mridulm replied:
If this is in executor, how are we ensuring shuffle is = 0, stage 1.0 and shuffle is = 0, stage 3.0 (2.0 computing for same shuffle is due to fetch failures) have unique id ?

ErikFang replied:
this issue had been handled with new solution as described in https://github.com/apache/incubator-celeborn/pull/1924#discussion_r1381725261

mridulm commented on client-spark/spark-3/src/main/scala/org/apache/spark/shuffle/celeborn/CelebornShuffleReader.scala:
```diff
@@ -114,7 +118,22 @@ class CelebornShuffleReader[K, C](
+                    -1,
+                    partitionId,
+                    SparkUtils.FETCH_FAILURE_ERROR_MSG,
+                    cioe)
```
Please add a note to `reportShuffleFetchFailure` that it must be a sync call.
Unlike in 'regular' spark codepaths, where `FetchFailedException` triggers special handling, here we are relying on `reportShuffleFetchFailure` to do the same.

ErikFang replied:
sure, will do

mridulm commented on client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:
```diff
@@ -75,6 +77,7 @@ class LifecycleManager(val appUniqueId: String, val conf: CelebornConf) extends
   val latestPartitionLocation =
     JavaUtils.newConcurrentHashMap[Int, ConcurrentHashMap[Int, PartitionLocation]]()
   private val userIdentifier: UserIdentifier = IdentityProvider.instantiate(conf).provide()
+  private val shuffleIdMapping = JavaUtils.newConcurrentHashMap[Int, ArrayBuffer[(Int, Boolean)]]()
```
Why not use:
```suggestion
  private val shuffleIdMapping = JavaUtils.newConcurrentHashMap[Int, scala.collection.mutable.LinkedHashMap[Int, Boolean]]()
```

This will avoid all the O(n) `find` scans through this - we can use `lastOption`

ErikFang replied:
by design, the max stage retry  number should be small, typically 3 - 4
Given a small size array, I suppose the O(N) lookup/update time should not be a performance issue, and I've seen too much HashMap/ConcurrentHashMap in LifecycleManager ...
I can accept the change if you insist

btw, lastOption is a good idea, an empty collection is unexpected, but it is always good to print a meaningful error message to users
will update

mridulm replied:
I was ofcourse assuming this (bounded size) is not the case - see my comments ([1](https://github.com/apache/incubator-celeborn/pull/1924#discussion_r1375372656), [2](https://github.com/apache/incubator-celeborn/pull/1924#discussion_r1375371317)).
'Typical' case should be small ofcourse !

ErikFang replied:
this issue had been handled with new solution as described in https://github.com/apache/incubator-celeborn/pull/1924#discussion_r1381725261

mridulm commented on client-spark/spark-3/src/main/scala/org/apache/spark/shuffle/celeborn/CelebornShuffleReader.scala:
```diff
@@ -60,6 +60,10 @@ class CelebornShuffleReader[K, C](
 
+    val shuffleId = shuffleClient.getShuffleId(handle.shuffleId)
+    logDebug(
+      s"get shuffleId $shuffleId for appShuffleId ${handle.shuffleId} attemptNum ${context.stageAttemptNumber()}")
```
In theory, there is a potential race condition here (task was earlier attempt is reading from newer shuffle id) - but should be beneign imo given it is for reads - writes is where it will be an issue (if code evolves to use `getShuffleId` for that).

ErikFang replied:
I agree
For DETERMINATE stage, read shuffle data from previous stage attempt will not introduce any issue
For INDETERMINATE stage, spark aborts the job now

mridulm reviewed (commented):
Left a few comments, thanks for fixing this !

ErikFang commented:
MapOutputTrackerMaster. unregisterAllMapAndMergeOutput() was added in spark 3.2 by [SPARK-32920](https://issues.apache.org/jira/browse/SPARK-32920), so checks with spark 3.0/3.1 fails

Can you suggest how should I handle this spark 3.x minor version issue in celeborn? @waitinfuture

mridulm commented:
`unregisterAllMapAndMergeOutput` was added when we contributed push based shuffle to Apache Spark.
It replaced the earlier `unregisterAllMapOutput` - which should work for 3.1 and older.

mridulm commented:
+CC @otterc as well.

waitinfuture commented:
Hi @mridulm , I still have a question about the deterministic. `DAGScheduler#submitMissingTasks` will only call `unregisterAllMapAndMergeOutput` if the current ShuffleMapStage is Indeterminate. What if the current stage is determinate, but its upstream stage is Indeterminate, and current stage throws `FetchFailed`, seems only its upstream
stage will be fully rerun, the current stage will only rerun its missing tasks. If this is the case, the current stage's output
can be incorrect.

mridulm commented:
Note that `RDD` can override `getOutputDeterministicLevel` - but the default behavior is what you described.

waitinfuture commented:
> Note that `RDD` can override `getOutputDeterministicLevel` - but the default behavior is what you described.

Yeah, so it can be that upstream stage is indeterminate while the downstream stage is determinate, and the incorrectness I described before can happen, right?

mridulm commented:
Top of my head, I cant think of a good scenario where prefix is indeterminate while suffix is determinate ... I was just mentioning from api pov ... For example, in the sample I had shared in dev@ list, I am forcing `DeterministicLevel` to be different :-)

waitinfuture commented:
> Top of my head, I cant think of a good scenario where prefix is indeterminate while suffix is determinate [1] ... I was just mentioning from api pov ... For example, in the sample I had shared in dev@ list, I am forcing `DeterministicLevel` to be different :-)
> 
> [1] Perhaps if all output is reduced to a single partition ? Not sure.

Single partition will not cause correctness issue anyway. User defining the deterministic level could lead to the issue, though I think it's not common.

mridulm commented:
Agree.
I was trying to come up with a (contrived) example where a user's RDD could override method - where the parent is indeterminate and yet rdd is determinate - this is purely thought exercise :-)

mridulm commented on client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SparkUtils.java:
```diff
@@ -92,6 +95,26 @@ public static String appUniqueId(SparkContext context) {
+              .append("-")
+              .append(context.stageAttemptNumber())
+              .toString();
+      return client.getShuffleId(handle.shuffleId(), appShuffleIdentifier, isWriter);
```
QQ: Do we want to cache this ? It should be immutable for a given shuffle id + stage id + attempt across all tasks.

waitinfuture replied:
+1

ErikFang replied:
caching shuffleid is a pretty good idea, however, I found it hard to implement in shuffleIdTracker.
shuffleIdTracker locates in SparkShuffleManager, which is not always true to have SparkShuffleManager when shuffleid is needed, for example

```
public class HashBasedShuffleWriterSuiteJ extends CelebornShuffleWriterSuiteBase {

  @Override
  protected ShuffleWriter<Integer, String> createShuffleWriter(
      CelebornShuffleHandle handle,
      TaskContext context,
      CelebornConf conf,
      ShuffleClient client,
      ShuffleWriteMetricsReporter metrics)
      throws IOException {
    return new HashBasedShuffleWriter<Integer, String, String>(
        SparkUtils.celebornShuffleId(client, handle, context, true),
        handle,
        context,
        conf,
        client,
        metrics,
        SendBufferPool.get(1, 30, 60));
  }
}
```

Currently, `shuffleIdCache` is added to ShuffleClientImpl, and works very similar to `reducePartitionMap` 
```
  // key: appShuffleIdentifier, value: shuffleId
  protected Map<String, Integer> shuffleIdCache = JavaUtils.newConcurrentHashMap();
```

mridulm commented on client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SparkShuffleManager.java:
```diff
@@ -79,6 +91,36 @@ public class SparkShuffleManager implements ShuffleManager {
+              new Function<Integer, HashSet<Integer>>() {
+                @Override
+                public HashSet<Integer> apply(Integer id) {
+                  return new HashSet<Integer>(shuffleId);
```
The constructor is specifying initial capacity.

```suggestion
                  return new HashSet<Integer>();
```

waitinfuture replied:
Right. I guess the intention is to put `shuffleId` when initialize the set, but since L110 always adds into the set, we don't need to add here.

ErikFang replied:
an unfortunate mistake...
I've fixed it, thanks!

mridulm commented on client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SparkShuffleManager.java:
```diff
@@ -79,6 +91,36 @@ public class SparkShuffleManager implements ShuffleManager {
+      HashSet<Integer> shuffleIds =
+          shuffleIdMap.computeIfAbsent(
+              appShuffleId,
+              new Function<Integer, HashSet<Integer>>() {
```
super nit: We can make it fairly succint with:
`shuffleIdMap.computeIfAbsent(appShuffleId, id -> new HashSet<>())` instead of `Function`

mridulm commented on client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SparkShuffleManager.java:
```diff
@@ -79,6 +91,36 @@ public class SparkShuffleManager implements ShuffleManager {
+
+    public void unregisterAppShuffleId(ShuffleClient shuffleClient, int appShuffleId) {
+      HashSet<Integer> shuffleIds = shuffleIdMap.remove(appShuffleId);
+      synchronized (shuffleIds) {
+        shuffleIds.forEach(shuffleId -> shuffleClient.cleanupShuffle(shuffleId));
+      }
```
In case of race or concurrent updates
```suggestion
      if (null != shuffleIds) {
        synchronized (shuffleIds) {
          shuffleIds.forEach(shuffleId -> shuffleClient.cleanupShuffle(shuffleId));
        }
      }  
```

ErikFang replied:
Agree, this is a good catch!

mridulm commented on client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SparkUtils.java:
```diff
@@ -220,4 +243,32 @@ public static <K, C> CelebornShuffleReader<K, C> createColumnarShuffleReader(
+    if (!UnregisterAllMapOutput_METHOD.isNoop()) {
+      UnregisterAllMapOutput_METHOD.bind(mapOutputTracker).invoke(shuffleId);
+      return;
+    }
```
Thanks for phrasing it this way, and trying both methods !
Internally, we have backported these changes - and so our internal 3.1 actually has `unregisterAllMapAndMergeOutput` :-)

mridulm reviewed (commented):
Just a few comments, mostly looks good to me.
I have not understood some of the Celeborn internal specific changes, but I am going to assume @waitinfuture and others will review that more thoroughly :)

otterc commented on client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:
```diff
@@ -564,6 +584,98 @@ class LifecycleManager(val appUniqueId: String, val conf: CelebornConf) extends
     commitManager.handleGetReducerFileGroup(context, shuffleId)
   }
 
+  private def handleGetShuffleIdForApp(
```
If we can add simple UTs for these changes, it will be great. It will encourage new contributors when they make changes to the same classes.

ErikFang replied:
sure, I was thinking about the same thing
currently, only spark integration tests are added, will check how LifecycleManager is tested now

ErikFang replied:
@waitinfuture can you help to provide a example for UT of LifecycleManager as reference?

waitinfuture replied:
> @waitinfuture can you help to provide a example for UT of LifecycleManager as reference?

You can refer to `WithShuffleClientSuite`

otterc reviewed (commented):
Thanks for this change.

SteNicholas commented on client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SparkShuffleManager.java:
```diff
@@ -35,9 +37,19 @@
 import org.apache.celeborn.common.util.ThreadUtils;
 import org.apache.celeborn.reflect.DynMethods;
 
+/*
```
```suggestion
/**
 * In order to support Spark stage resubmission when ShuffleReader's fetch fails, Celeborn's shuffle id should be distinguished from Spark's shuffle id.
 * Spark's shuffle id is assigned when ShuffleDependency is constructed. Meanwhile, all attempts of a Spark Stage have the same shuffle id.
 * When ShuffleReader fails to fetch shuffle data from worker and throws {@link FetchFailedException}, Spark DAGScheduler resubmits the failed ResultStage and corresponding ShuffleMapStage.
 * But Celeborn couldn't make shuffle data any different from previous failed or resubmitted ShuffleMapStage with the same shuffle id.
 * Celeborn takes retry of stage in consideration in current implementation. LifecycleManager generates and tracks usage of shuffle id for Spark and Celeborn.
 * Then, Spark shuffle Reader/Writer gets shuffle id from LifecycleManager with GetShuffleId request.
 */
```

SteNicholas commented on client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SparkShuffleManager.java:
```diff
@@ -125,6 +167,13 @@ private void initializeLifecycleManager() {
           lifecycleManager = new LifecycleManager(appUniqueId, celebornConf);
+          if (celebornConf.clientFetchThrowsFetchFailure()) {
+            MapOutputTrackerMaster mapOutputTracker =
+                (MapOutputTrackerMaster) SparkEnv.get().mapOutputTracker();
```
Why convert to `MapOutputTrackerMaster`?

ErikFang replied:
this function is executed in spark driver that has a MapOutputTrackerMaster, and the method unregisterAllMapAndMergeOutput is a member of MapOutputTrackerMaster not MapOutputTracker

SteNicholas commented on client-spark/spark-3/src/main/scala/org/apache/spark/shuffle/celeborn/CelebornShuffleReader.scala:
```diff
@@ -124,12 +146,31 @@ class CelebornShuffleReader[K, C](
+      try {
+        iter
+      } catch {
+        case e: CelebornIOException =>
```
How to handle other kind of Exception?

ErikFang replied:
I've discussed with @waitinfuture previously, and was told most exception is converted to CelebornIOException before throwing it.
It should be safe in current implementation

SteNicholas commented on client/src/main/java/org/apache/celeborn/client/ShuffleClient.java:
```diff
@@ -212,4 +212,13 @@ public abstract ConcurrentHashMap<Integer, PartitionLocation> getPartitionLocati
+
+  public abstract int getShuffleId(int appShuffleId, String appShuffleIdentifier, boolean isWriter);
+
+  /**
```
```suggestion
  /**
   * Reports the failure of shuffle data fetch given shuffle id of Spark and Celeborn.
   */
```

SteNicholas commented on common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:
```diff
@@ -3207,6 +3208,15 @@ object CelebornConf extends Logging {
       .createWithDefault(3)
 
+  val CLIENT_FETCH_THROWS_FETCH_FAILURE: ConfigEntry[Boolean] =
+    buildConf("celeborn.client.fetch.throwsFetchFailure")
```
Is the config option suitable for all engine client?

ErikFang replied:
changes to "celeborn.client.spark.fetch.throwsFetchFailure"

waitinfuture commented on common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:
```diff
@@ -3207,6 +3208,15 @@ object CelebornConf extends Logging {
 
+  val CLIENT_FETCH_THROWS_FETCH_FAILURE: ConfigEntry[Boolean] =
+    buildConf("celeborn.client.fetch.throwsFetchFailure")
+      .withAlternative("celeborn.fetch.throwsFetchFailure")
```
`withAlternative` is used for backward compatibility. Since this is a new config, we don't need to call `withAlternative`

waitinfuture commented on client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SparkUtils.java:
```diff
@@ -220,4 +243,32 @@ public static <K, C> CelebornShuffleReader<K, C> createColumnarShuffleReader(
+          .orNoop()
+          .build();
+
+  // for spark 3.1 and 2.4, see detail in SPARK-32920
```
Since this class is in spark-3 package, we don't need to consider spark2.x here

waitinfuture commented on client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SparkUtils.java:
```diff
@@ -92,6 +95,26 @@ public static String appUniqueId(SparkContext context) {
         .getOrElse(context::applicationId);
   }
 
+  public static int celebornShuffleId(
```
IMO it's better to move this method to `ShuffleClient` for potentially other engines. We can change the argument list to `appShuffleId, appShuffleIdentifier, isWriter`.

ErikFang replied:
`public abstract int getShuffleId(int appShuffleId, String appShuffleIdentifier, boolean isWriter);`
was already added to ShuffleClient 

I prefer to keep SparkUtils.celebornShuffleId() method as a place to check whether it needs to get shuffleId from LifecycleManager
```
  public static int celebornShuffleId(
      ShuffleClient client,
      CelebornShuffleHandle<?, ?, ?> handle,
      TaskContext context,
      Boolean isWriter) {
    if (handle.throwsFetchFailure()) {
      String appShuffleIdentifier = getAppShuffleIdentifier(handle.shuffleId(), context);
      return client.getShuffleId(handle.shuffleId(), appShuffleIdentifier, isWriter);
    } else {
      return handle.shuffleId();
    }
  }
```
otherwise, it would be hard to implement the check/fallback logic in place like constructor
```
  public SortBasedShuffleWriter(
      CelebornShuffleHandle<K, V, C> handle,
      TaskContext taskContext,
      CelebornConf conf,
      ShuffleClient client,
      ShuffleWriteMetricsReporter metrics,
      ExecutorService executorService,
      SendBufferPool sendBufferPool)
      throws IOException {
    this(
        SparkUtils.celebornShuffleId(client, handle, taskContext, true),
        handle.dependency(),
        handle.numMappers(),
        taskContext,
        conf,
        client,
        metrics,
        executorService,
        sendBufferPool);
  }
```

waitinfuture replied:
We can add `shuffleId` as an argument to the constructor, as `HashBasedShuffleWriter` does. `SparkUtils.celebornShuffleId()` is a small wrapper for `ShuffleClient.getShuffleId`. I'm fine with this anyway.

waitinfuture reviewed (commented):
Current code mostly LGTM except some small comments. Thanks!

waitinfuture commented on client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SparkShuffleManager.java:
```diff
@@ -217,10 +267,14 @@ public <K, V> ShuffleWriter<K, V> getWriter(
                 h.lifecycleManagerPort(),
                 celebornConf,
                 h.userIdentifier());
+        int shuffleId = SparkUtils.celebornShuffleId(shuffleClient, h, context, true);
```
Right, maybe we should also change the constructor of `ColumnarHashBasedShuffleWriter` to add `shuffleId` in the argument list.

ErikFang replied:
this issue has been fixed in latest PR

ErikFang commented on client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:
```diff
@@ -556,6 +566,59 @@ class LifecycleManager(val appUniqueId: String, val conf: CelebornConf) extends
+      }
+    }
+
+    val shuffleId = shuffleIdMapping.get(appShuffleId).last._1
```
the issue doesn't exists in latest code

ErikFang commented on client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:
```diff
@@ -564,6 +584,98 @@ class LifecycleManager(val appUniqueId: String, val conf: CelebornConf) extends
+                  ret = false
+              }
+              shuffleIds.put(appShuffleIdentifier, (shuffleId, false))
+              unregisterShuffle(shuffleId)
```
I suppose other 900 tasks fails with fetch failure won't change any thing in current design
anyway, I don't have any problem to delay the unregister shuffle, will change it

waitinfuture replied:
If the reduce stage is determinate, then if the 900 reduce tasks succeed, they will not be rerun.

ErikFang commented on client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SparkShuffleManager.java:
```diff
@@ -127,11 +140,18 @@ private void initializeLifecycleManager() {
 
+  private void registerCallback() {
+    MapOutputTrackerMaster mapOutputTracker = (MapOutputTrackerMaster) SparkEnv.get().mapOutputTracker();
+    Consumer<Integer> c =  (shuffleId) -> mapOutputTracker.unregisterAllMapAndMergeOutput(shuffleId);
```
https://lists.apache.org/thread/t8z8pc2dw6ooclk57y4sn1loof68tthq
Thanks for @mridulm to give a detailed explanation in mail list

SteNicholas commented on common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:
```diff
@@ -3228,6 +3229,14 @@ object CelebornConf extends Logging {
+  val CLIENT_FETCH_THROWS_FETCH_FAILURE: ConfigEntry[Boolean] =
+    buildConf("celeborn.client.spark.fetch.throwsFetchFailure")
+      .categories("client")
+      .version("0.4.0")
```
```suggestion
      .version("0.3.2")
```

waitinfuture replied:
Hi @SteNicholas , since this is a big feature, I suggest not merge to branch-0.3

SteNicholas commented on client-spark/common/src/main/java/org/apache/spark/shuffle/celeborn/ExecutorShuffleIdTracker.java:
```diff
@@ -0,0 +1,47 @@
+    HashSet<Integer> shuffleIds = shuffleIdMap.remove(appShuffleId);
+    if (shuffleIds != null) {
+      synchronized (shuffleIds) {
+        shuffleIds.forEach(shuffleId -> shuffleClient.cleanupShuffle(shuffleId));
```
```suggestion
        shuffleIds.forEach(shuffleClient::cleanupShuffle);
```

ErikFang replied:
fixed

waitinfuture reviewed (approved):
The current code LGTM, I tested without replication and kill worker in either write stage and read stage, results are good. cc @mridulm @otterc do you have more comments on this PR?

otterc reviewed (approved):
Just have few nits

waitinfuture commented:
Here is my test. Note: should first cherry-pick another fix: https://github.com/apache/incubator-celeborn/pull/2104
Command:
```
spark-shell --conf spark.executor.instances=100 --conf spark.shuffle.manager=org.apache.spark.shuffle.celeborn.SparkShuffleManager --conf spark.serializer=org.apache.spark.serializer.KryoSerializer --conf spark.celeborn.master.endpoints=master-1-1:9097 --conf spark.shuffle.service.enabled=false --conf spark.celeborn.client.push.replicate.enabled=false --conf spark.celeborn.client.spark.fetch.throwsFetchFailure=true
```

Test case:
```
spark.sparkContext.parallelize(1 to 1000, 1000).flatMap( _ => (1 to 1000000).iterator.map(num => num)).repartition(1000).count
```

When I killed a worker during shuffle write stage:
![image](https://github.com/apache/incubator-celeborn/assets/948245/27b7aee4-ed4f-45a7-840b-8b27a5866ffb)

When I killed a worker during shuffle read stage:
![image](https://github.com/apache/incubator-celeborn/assets/948245/b0c39c34-f64d-445e-acaf-669fa5e63bcf)

I'm going to test TPCDS.

mridulm reviewed (approved):
Just some nits and queries. Thanks for working on this @ErikFang, excited to see this as part of next release :-)

waitinfuture commented:
There is another issue to be solved. Say a determinate shuffleMapStage A also generates shuffle out, at some time it finishes 100 tasks, with another 100 tasks to be run, then it encounters fetch failure. It triggers upstream stage rerun successfully and went back. At this time, since it uses a new stage attemptid, it has a new shuffleIdentifier, so it creates a new shuffle id. Since it's determinate, it only runs the remaining 100 tasks. However, Celeborn only trigger StageEnd when it receives MapperEnd for `number of partitions of this stage`, which is 200, so it will not trigger StageEnd. But Spark thinks all 200 tasks are successful, so it schedules the downstream stage, the downstream stage can't find a finished shuffleId, and finally causes job failure.

IMO we need to get the deterministic information and only create a new shuffleId for a writer when it's nondeterminate,  WDYT @ErikFang  @mridulm ?

ErikFang commented:
> There is another issue to be solved. Say a determinate shuffleMapStage A also generates shuffle out, at some time it finishes 100 tasks, with another 100 tasks to be run, then it encounters fetch failure. It triggers upstream stage rerun successfully and went back. At this time, since it uses a new stage attemptid, it has a new shuffleIdentifier, so it creates a new shuffle id. Since it's determinate, it only runs the remaining 100 tasks. However, Celeborn only trigger StageEnd for a shuffleId when it receives MapperEnd for `number of partitions of this stage`, which is 200, so it will not trigger StageEnd. But Spark thinks all 200 tasks are successful, so it schedules the downstream stage, the downstream stage can't find a finished shuffleId, and finally causes job failure.
> 
> IMO we need to get the deterministic information and only create a new shuffleId for a writer when it's nondeterminate, WDYT @ErikFang @mridulm ?

I suppose it is doable

https://github.com/apache/spark/blob/v3.2.0/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala#L468
DAGScheduler creates ShuffleMapStage with rdd obtained from ShuffleDependency, and ShuffleManager gets ShuffleDependency object in registerShuffle(...), then we can have registerShuffle(...) to pass (appShuffleId, rdd.outputDeterministicLevel == INDETERMINATE) to LifeCycleManager

waitinfuture commented:
I tested with the latest code (43e4ee22a3e04caf594cc8d588d068ba1617c7f3) together with CELEBORN-1130, and I'm good with the result.

I tested TPCDS q11 1T with 8 workers and replication off. I successively killed 6 workers out of 8, and the query succeeds with correct result (I set `spark.stage.maxConsecutiveAttempts=100`). We can see lots of retry stages:
![image](https://github.com/apache/incubator-celeborn/assets/948245/af4fefd9-43b5-4160-ab19-5ebb0ec6438b)

@SteNicholas please rebase the lates code, I'm happy to merge this PR :)
cc @mridulm @otterc @SteNicholas @RexXiong

otterc commented on client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:
```diff
@@ -1331,6 +1352,10 @@ class LifecycleManager(val appUniqueId: String, val conf: CelebornConf) extends
     appShuffleTrackerCallback = Some(callback)
   }
 
+  def registerAppShuffleDeterminate(appShuffleId: Int, determinate: Boolean): Unit = {
```
Nit: `determinate` is an argument to the method, so shouldn't we just call this method `registerAppShuffle`

ErikFang replied:
Per my perspective, the whole point of this method is to pass determinate information of ShuffleMapStage to LifecycleManager, so it is a better name than registerAppShuffle
actually, it is handleGetShuffleIdForApp method that handles the job of "registerAppShuffle"

otterc commented on client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:
```diff
@@ -79,6 +81,13 @@ class LifecycleManager(val appUniqueId: String, val conf: CelebornConf) extends
+    scala.collection.mutable.LinkedHashMap[String, (Int, Boolean)]]()
+  private val shuffleIdGenerator = new AtomicInteger(0)
+  // app shuffle id -> whether shuffle is determinate, rerun of a indeterminate shuffle gets different result
+  private val appShuffleDeterminateMap = JavaUtils.newConcurrentHashMap[Int, Boolean]();
```
Do we need to maintain this state? Can the `GetShuffleId` message not include whether the shuffle is determinate or not?

ErikFang replied:
I like the idea of reducing the status~
however, it seems hard to get rdd/stage DeterministicLevel in task/executor, it is where GetShuffleId RPC is called

ErikFang commented on client-spark/spark-2/src/main/java/org/apache/spark/shuffle/celeborn/SparkShuffleManager.java:
```diff
@@ -105,6 +107,12 @@ private void initializeLifecycleManager(String appId) {
+                (MapOutputTrackerMaster) SparkEnv.get().mapOutputTracker();
+            lifecycleManager.registerShuffleTrackerCallback(
+                shuffleId -> mapOutputTracker.unregisterAllMapOutput(shuffleId));
+          }
```
yeah, I think the MT concerns are reasonable, but I'm afraid changing the constructor signature may affect other engines
how about adding @volatile to appShuffleTrackerCallback?

SteNicholas commented:
@waitinfuture, @ErikFang, @mridulm, @otterc, `shuffleClient#readPartition` in `LifecycleManager` should also throw fetch failure for `CelebornIOException` and `PartitionUnRetryAbleException`. I have fixed this bug internal and the test result that `shuffleClient#readPartition` throws fetch failure is successful, which is as follows:
![wecom-temp-164223-b36261fd0b10d3bacdaef3928a45efbb](https://github.com/apache/incubator-celeborn/assets/10048174/2134e972-d4f1-40b8-bf17-1e075d4dec8a)
<img width="1398" alt="image" src="https://github.com/apache/incubator-celeborn/assets/10048174/14750e2b-42ff-42b0-9896-a2d377be9182">
![wecom-temp-82543-e6d89ca9cb6caa6f63f2e73dffb70aeb](https://github.com/apache/incubator-celeborn/assets/10048174/0a62d1fd-cae2-42d8-bddb-80d7d18fe0b6)

ErikFang commented on client-spark/spark-2/src/main/scala/org/apache/spark/shuffle/celeborn/CelebornShuffleReader.scala:
```diff
@@ -59,6 +60,11 @@ class CelebornShuffleReader[K, C](
     val serializerInstance = dep.serializer.newInstance()
 
+    val shuffleId = SparkUtils.celebornShuffleId(shuffleClient, handle, context, false)
+    shuffleIdTracker.track(handle.shuffleId, shuffleId)
```
https://github.com/apache/incubator-celeborn/blob/v0.3.1-incubating/client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java#L1521C6-L1521C6

in ShuffleClientImpl.cleanupShuffle(), reduceFileGroupsMap is cleaned, which is used for reads

ErikFang commented on tests/spark-it/src/test/scala/org/apache/celeborn/tests/spark/CelebornFetchFailureSuite.scala:
```diff
@@ -0,0 +1,193 @@
+      .map { i => (i, value) }.groupByKey(16).collect()
+
+    // verify result
+    assert(hook.executed.get() == true)
```
I prefer to be explicit

ErikFang commented on client-spark/spark-2/src/test/java/org/apache/spark/shuffle/celeborn/HookedCelebornShuffleManager.java:
```diff
@@ -0,0 +1,45 @@
+
+public class HookedCelebornShuffleManager extends SparkShuffleManager {
+
+  private static ShuffleManagerHook shuffleReaderGetHook = null;
```
ShuffleManagerHook is not a shuffleReader, but a hook

ErikFang commented on client-spark/spark-2/src/test/java/org/apache/spark/shuffle/celeborn/HookedCelebornShuffleManager.java:
```diff
@@ -0,0 +1,45 @@
+import org.apache.spark.shuffle.ShuffleHandle;
+import org.apache.spark.shuffle.ShuffleReader;
+
+public class HookedCelebornShuffleManager extends SparkShuffleManager {
```
renamed to TestCelebornShuffleManager in latest commit

ErikFang commented on tests/spark-it/src/test/scala/org/apache/celeborn/tests/spark/CelebornFetchFailureSuite.scala:
```diff
@@ -0,0 +1,193 @@
+
+    hook.executed.set(true)
+
+    val x1 = rdd1.count()
```
fixed

ErikFang commented on tests/spark-it/src/test/scala/org/apache/celeborn/tests/spark/CelebornFetchFailureSuite.scala:
```diff
@@ -0,0 +1,193 @@
+
+    val celebornConf = SparkUtils.fromSparkConf(sparkSession.sparkContext.getConf)
+    val hook = new ShuffleReaderGetHook(celebornConf)
+    HookedCelebornShuffleManager.registerReaderGetHook(hook);
```
fixed

ErikFang commented:
rebased the code against latest main branch, renamed HookedCelebornShuffleManager to TestCelebornShuffleManager, added shuffle id with FETCH_FAILURE_ERROR_MSG to help debug, decorated appShuffleTrackerCallback with volatile

I favor adding NoOpShuffleIdTracker and UT of PbGetShuffleId/PbReportShuffleFetchFailure message handing in a follow-up PR
gentle ping @waitinfuture to take a look

waitinfuture commented on client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:
```diff
@@ -628,6 +648,118 @@ class LifecycleManager(val appUniqueId: String, val conf: CelebornConf) extends
+                  None
+
+              val shuffleId: Integer =
+                if (determinate && candidateShuffle.isDefined) {
```
nit: no need to check `determinate` here

waitinfuture reviewed (approved):
LGTM, thanks! Merging to main(v0.4.0). Thanks all reviewers, especially @mridulm for giving valuable insights of spark's rerun internals!

s0nskar commented on client-spark/common/src/main/java/org/apache/spark/shuffle/celeborn/ExecutorShuffleIdTracker.java:
```diff
@@ -0,0 +1,47 @@
+import org.apache.celeborn.common.util.JavaUtils;
+
+public class ExecutorShuffleIdTracker {
+  // track appShuffleId -> shuffleId Set in executor for cleanup
```
shuffleId here is celebornShuffleId?
